<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных-и-изучение-общей-информации" data-toc-modified-id="Загрузка-данных-и-изучение-общей-информации-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных и изучение общей информации</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Обработка-пропусков" data-toc-modified-id="Обработка-пропусков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Обработка пропусков</a></span></li><li><span><a href="#Замена-типа-данных" data-toc-modified-id="Замена-типа-данных-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Замена типа данных</a></span></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Обработка дубликатов</a></span></li><li><span><a href="#Лемматизация" data-toc-modified-id="Лемматизация-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Лемматизация</a></span></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Категоризация данных</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Связь-присущих-заемщикам-признаков-и-возврата-кредита-в-срок" data-toc-modified-id="Связь-присущих-заемщикам-признаков-и-возврата-кредита-в-срок-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Связь присущих заемщикам признаков и возврата кредита в срок</a></span><ul class="toc-item"><li><span><a href="#Есть-ли-зависимость-между-наличием-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-наличием-детей-и-возвратом-кредита-в-срок?-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Есть ли зависимость между наличием детей и возвратом кредита в срок?</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.3.1"><span class="toc-item-num">3.3.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Как разные цели кредита влияют на его возврат в срок?</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.4.1"><span class="toc-item-num">3.4.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

# Выявление факторов, влияющих на возврат заемщиками кредита в срок

Для кредитного отдела банка необходимо выяснить, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования нужны для построения модели кредитного скоринга.

Для анализа предоставлена таблица с информацией о клиентах.

**Описание данных (содержание столбцов таблицы)**

- `children` — количество детей в семье;

- `days_employed` — общий трудовой стаж в днях;

- `dob_years` — возраст клиента в годах;

- `education` — уровень образования клиента;

- `education_id` — идентификатор уровня образования;

- `family_status` — семейное положение;

- `family_status_id` — идентификатор семейного положения;

- `gender` — пол клиента;

- `income_type` — тип занятости;

- `debt` — имел ли задолженность по возврату кредитов;

- `total_income` — ежемесячный доход;

- `purpose` — цель получения кредита.

## Загрузка данных и изучение общей информации

In [1]:
import pandas as pd
import warnings
from IPython.display import display
from collections import Counter
from pymystem3 import Mystem

In [2]:
warnings.filterwarnings('ignore') # отключение показа предупреждения, связанного с библиотекой pymystem3

In [3]:
data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv') #запишем данные в переменную

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 1.6+ MB


In [5]:
data.shape

(21525, 12)

In [6]:
data.head(5)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

### Вывод

Датафрейм содержит 21525 строк и 12 столбцов.

Обращают на себя внимание пропущенные значения в столбцах `days_employed` и `total_income`, а также разный регистр букв в строковых значениях столбца `education`, а также отрицательные значения в столбце `days_employed`, дублирующие друг друга по смыслу столбцы `education` и `education_id`, `family_status` и `family_status_id`.

## Предобработка данных

### Обработка пропусков

In [7]:
print('Доля пропущенных значений в столбце days_employed составляет {:.2%}'.format(data['days_employed'].isna().sum() / 21252))
print('---' * 15)
print('Минимальное значение days_employed:', data['days_employed'].min())
print('Максимальное значение days_employed:', data['days_employed'].max())

Доля пропущенных значений в столбце days_employed составляет 10.23%
---------------------------------------------
Минимальное значение days_employed: -18388.949900568383
Максимальное значение days_employed: 401755.40047533


In [8]:
# пропуски помечаются как 0\1 и среднее покажет долю 1 (1 там, где пропуск)
# выберем только те, где есть пропуски
data.isna().mean()[data.isna().mean() != 0]

days_employed    0.100999
total_income     0.100999
dtype: float64

Убедимся, что клиенты с пропущенными значениями не безработные.

In [9]:
print(data['days_employed'].isna().sum()) #количество пропусков в days_employed
display(data.loc[(data['days_employed'].isna()) & data['total_income'].isna()]['income_type'])

2174


12         пенсионер
26       госслужащий
29         пенсионер
41       госслужащий
55         пенсионер
            ...     
21489      компаньон
21495      сотрудник
21497      компаньон
21502      сотрудник
21510      сотрудник
Name: income_type, Length: 2174, dtype: object

Мы убедились не только в том, что клиенты с пропущенными значениями не безработные, но и в том, что там, где пропущены значения в days_employed, там же пропущены значения в total_income.

In [10]:
#заменим пропуски, но сначала приведем данные к единому "формату" по времени
print('401755 дней в годах составляет {:d} лет'.format(int(401755 / 365)))
print('18388 дней в годах составляет {:d} лет'.format(int(18388 / 365)))

401755 дней в годах составляет 1100 лет
18388 дней в годах составляет 50 лет


In [11]:
def to_positive_in_days(value): #функция для приведения значений к единому формату
    if value < 0:
        return -value #если значение отрицательное, вернуть модуль
    else:
        return (value / 24) #в ином случае разделить значение на 24 (дни работы = часы работы / количество часов в сутках)

    
data['days_employed'] = data['days_employed'].apply(to_positive_in_days) #применение функции к столбцу

median_days_employed = data['days_employed'].median()

In [12]:
print('Минимальное значение days_employed:', data['days_employed'].min())
print('Максимальное значение days_employed:', data['days_employed'].max())
print('Среднее значение days_employed:', data['days_employed'].mean())
print('Медиана days_employed:', median_days_employed)

Минимальное значение days_employed: 24.14163324048118
Максимальное значение days_employed: 18388.949900568383
Среднее значение days_employed: 4641.641176180656
Медиана days_employed: 2194.220566878695


In [13]:
data['days_employed'] = data['days_employed'].fillna(median_days_employed) #замена пропусков days_employed медианным значением

print('Пропущенных значений в столбце days_employed:', data['days_employed'].isnull().sum())

Пропущенных значений в столбце days_employed: 0


In [14]:
print('Минимальное значение total_income:', data['total_income'].min())
print('Максимальное значение total_income:', data['total_income'].max())
print('Среднее значение total_income:', data['total_income'].mean())
median_total_income = data['total_income'].median()
print('Медиана total_income:', median_total_income) 

Минимальное значение total_income: 20667.26379327158
Максимальное значение total_income: 2265604.028722744
Среднее значение total_income: 167422.30220817294
Медиана total_income: 145017.93753253992


In [15]:
#рассмотрим медиану total_income по группам типа занятости
display(data.groupby('income_type')['total_income'].median())

income_type
безработный        131339.751676
в декрете           53829.130729
госслужащий        150447.935283
компаньон          172357.950966
пенсионер          118514.486412
предприниматель    499163.144947
сотрудник          142594.396847
студент             98201.625314
Name: total_income, dtype: float64

In [16]:
#записываем в переменную категории типа занятости клиентов с пропущенными значениями дохода
nan_income_unique = data.loc[data['total_income'].isna()]['income_type'].unique()
for nan_income in nan_income_unique: #заменяем пропущенные значения медианой дохода по типу занятости
    type_income_median = data.loc[data['income_type'] == nan_income, 'total_income'].median()
    data.loc[data['income_type'] == nan_income, 'total_income'] = data.loc[data['income_type'] == nan_income, 'total_income'].fillna(type_income_median) 

In [17]:
print('Пропущенных значений в столбце total_income:', data['total_income'].isnull().sum())

Пропущенных значений в столбце total_income: 0


In [18]:
data.info() #пропущенные значения отсутствуют

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 1.6+ MB


In [19]:
#проверим другие количественные данные на наличие нестандартных вариантов (например, пропущенное значение возраста как 0)
print('Значения количества детей: MIN {} | MAX {}'.format(data['children'].min(), data['children'].max()))
print('Значения возраста клиентов: MIN {} | MAX {}'.format(data['dob_years'].min(), data['dob_years'].max()))

Значения количества детей: MIN -1 | MAX 20
Значения возраста клиентов: MIN 0 | MAX 75


In [20]:
#рассмотрим подробнее значения количества детей
print('Уникальные значения количества детей:', data['children'].sort_values().unique())

Уникальные значения количества детей: [-1  0  1  2  3  4  5 20]


In [21]:
#между 5 и 20 промежуточных значений нет, очевидно, это опечатка, как и -1
data.loc[data['children'] == -1, 'children'] = 1
data.loc[data['children'] == 20, 'children'] = 2

In [22]:
print('Значения количества детей: MIN {} | MAX {}'.format(data['children'].min(), data['children'].max()))
print('Уникальные значения количества детей:', data['children'].sort_values().unique())

Значения количества детей: MIN 0 | MAX 5
Уникальные значения количества детей: [0 1 2 3 4 5]


Подробнее рассмотрим значения возраста клиентов:

In [23]:
display(data.groupby('dob_years')['dob_years'].count().head(5)) #значение возраста 101 клиента составляет 0

dob_years
0     101
19     14
20     51
21    111
22    183
Name: dob_years, dtype: int64

Заполним нулевые значения возраста в зависимости от типа занятости клиента (пенсионер или не пенсионер):

In [24]:
#выбрали средние значения, так как сильно отклоняющихся значений в возрасте не будет
#методом astype заранее переводим float в int
senior_age_mean = data.loc[(data['income_type'] == 'пенсионер') & (data['dob_years'] != 0), 'dob_years'].mean().astype('int')
non_senior_age_mean = data.loc[(data['income_type'] != 'пенсионер') & (data['dob_years'] != 0), 'dob_years'].mean().astype('int')
data.loc[(data['dob_years'] == 0) & (data['income_type'] == 'пенсионер'), 'dob_years'] = senior_age_mean
data.loc[(data['dob_years'] == 0) & (data['income_type'] != 'пенсионер'), 'dob_years'] = non_senior_age_mean

In [25]:
display(data.groupby('dob_years')['dob_years'].count().head(5))

dob_years
19     14
20     51
21    111
22    183
23    254
Name: dob_years, dtype: int64

Проверим строковые категориальные данные на наличие значений типа прочерков, "None", "N/A" и т.п.

In [26]:
category_columns = ['education', 'family_status', 'gender', 'income_type', 'purpose']
for column in category_columns:
    print(column)
    print(data[column].unique())
    print('---' * 15)

education
['высшее' 'среднее' 'Среднее' 'СРЕДНЕЕ' 'ВЫСШЕЕ' 'неоконченное высшее'
 'начальное' 'Высшее' 'НЕОКОНЧЕННОЕ ВЫСШЕЕ' 'Неоконченное высшее'
 'НАЧАЛЬНОЕ' 'Начальное' 'Ученая степень' 'УЧЕНАЯ СТЕПЕНЬ'
 'ученая степень']
---------------------------------------------
family_status
['женат / замужем' 'гражданский брак' 'вдовец / вдова' 'в разводе'
 'Не женат / не замужем']
---------------------------------------------
gender
['F' 'M' 'XNA']
---------------------------------------------
income_type
['сотрудник' 'пенсионер' 'компаньон' 'госслужащий' 'безработный'
 'предприниматель' 'студент' 'в декрете']
---------------------------------------------
purpose
['покупка жилья' 'приобретение автомобиля' 'дополнительное образование'
 'сыграть свадьбу' 'операции с жильем' 'образование'
 'на проведение свадьбы' 'покупка жилья для семьи' 'покупка недвижимости'
 'покупка коммерческой недвижимости' 'покупка жилой недвижимости'
 'строительство собственной недвижимости' 'недвижимость'
 'строительс

Просмотрев уникальные значения выше, видим 'XNA' в значениях пола клиентов. Рассмотрим данные об этом клиенте подробнее


In [27]:
data.loc[data['gender'] == 'XNA']

children  days_employed  dob_years            education  education_id  \
10701         0    2358.600502         24  неоконченное высшее             2   

          family_status  family_status_id gender income_type  debt  \
10701  гражданский брак                 1    XNA   компаньон     0   

        total_income               purpose  
10701  203905.157261  покупка недвижимости

Попробуем установить неиболее вероятный пол клиента, используя значение типа занятости.

In [28]:
data.loc[data['income_type'] == 'компаньон', 'gender'].value_counts()

F      3197
M      1887
XNA       1
Name: gender, dtype: int64

In [29]:
#компаньонами в 1,7 раза чаще выступают женщины, поэтому более вероятно, что клиент XNA женского пола
data.loc[data['gender'] == 'XNA', 'gender'] = 'F'

In [30]:
#проверяем уникальные значения пола
print(data['gender'].unique())

['F' 'M']


**Вывод**

Мы проверили методом isna() наличие пропущенных значений, а также просмотрели уникальные значения категориальных данных, чтобы найти среди них "нереальные" (например, возраст клиентов 0) либо пропуски, обозначенные строкой (пол 'XNA').

Таким образом, были обнаружены пропущенные значения в виде NaN (для стажа и доходов), а также в форматах int (возраст) и строковых значений (пол). При этом попутно были обнаружены и исправлены некорректные данные (отрицательный  и "тысячелетний" стаж, количество детей).

Удалив все строки, в которых встречается NaN, мы бы лишили себя почти 10% данных, что негативно отразится на выводах. Поэтому пропуски должны были быть заполнены.

Для заполнения пропусков в стаже использована медиана трудового стажа по всем клиентам и метод fillna().  Для заполнения пропусков дохода использована медиана дохода по категории, к которой относится клиент. Среднее значение для заполнения пропусков не использовалось, так как на сильно него влияют отклоняющиеся значения.

Пропуск пола был заполнен на основании типа занятости клиента (тип занятости клиента более распространен среди женщин).

Нулевые значения возраста были заполнены на основании того, относится или нет клиент к пенсионерам, для заполнения использованы средние значения (так как сильно отклоняющиеся значения в возрасте встретить сомнительно).

Возможные причины появления пропусков данных:

- данные по случайности не были внесены
- данные были намеренно сокрыты
- значение не существует, поэтому его нельзя внести
- технический сбой
- удаление уже внесенных данных

### Замена типа данных

In [31]:
data['days_employed'] = data['days_employed'].astype('int')
data['total_income'] = data['total_income'].astype('int')

In [32]:
data.info() #проверка замены типа

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21525 non-null  int64 
 1   days_employed     21525 non-null  int32 
 2   dob_years         21525 non-null  int64 
 3   education         21525 non-null  object
 4   education_id      21525 non-null  int64 
 5   family_status     21525 non-null  object
 6   family_status_id  21525 non-null  int64 
 7   gender            21525 non-null  object
 8   income_type       21525 non-null  object
 9   debt              21525 non-null  int64 
 10  total_income      21525 non-null  int32 
 11  purpose           21525 non-null  object
dtypes: int32(2), int64(5), object(5)
memory usage: 1.4+ MB


**Вывод**

На этапе изучения информации о файле было установлено, что вещественный тип данных float имеют столбцы days_employed И total_income.

Для изменения типа данных использован метод astype() с аргументом 'int'. Метод to_numeric() для данной задачи не подходит, так как преобразовывает данные в вещественный тип float, тогда как нам необходимо было преобразование из вещественного в целочисленный int.

### Обработка дубликатов

In [33]:
#перед поиском и удалением дубликатов приведем строковые значения education к единому регистру
data['education'] = data['education'].str.lower()

In [34]:
print(data['education'].unique()) #проверяем уникальные значения family_status

['высшее' 'среднее' 'неоконченное высшее' 'начальное' 'ученая степень']


In [35]:
data.duplicated().sum() #поиск дубликатов, их количество

71

In [36]:
data = data.drop_duplicates().reset_index(drop=True) #удаление дубликатов

In [37]:
data.duplicated().sum() #проверка наличия дубликатов

0

**Вывод**

Перед поиском дубликатов значения столбца education (в нем на более ранних этапах проекта выявлены одинаковые значения в разном регистре) приведены к нижнему регистру цепочкой методов str.lower(), после чего методом duplicated() выявлено наличие дубликатов, которые были удалены методом drop_duplicates() в связке с методом reset_index() (сброс индекса).

Возможные причины появления дубликатов в данных:
- случайный повторный ввод данных
- ввод одинаковых данных со строковыми значениями в разном регистре (отсутствие единой формы для значений)
- технический сбой
- ошибки при копировании данных

### Лемматизация

In [38]:
m = Mystem() 

In [39]:
unique_purposes = data['purpose'].unique()

Лемматизируем только уникальные цели получения кредита.

In [40]:
print(unique_purposes)

['покупка жилья' 'приобретение автомобиля' 'дополнительное образование'
 'сыграть свадьбу' 'операции с жильем' 'образование'
 'на проведение свадьбы' 'покупка жилья для семьи' 'покупка недвижимости'
 'покупка коммерческой недвижимости' 'покупка жилой недвижимости'
 'строительство собственной недвижимости' 'недвижимость'
 'строительство недвижимости' 'на покупку подержанного автомобиля'
 'на покупку своего автомобиля' 'операции с коммерческой недвижимостью'
 'строительство жилой недвижимости' 'жилье'
 'операции со своей недвижимостью' 'автомобили' 'заняться образованием'
 'сделка с подержанным автомобилем' 'получение образования' 'автомобиль'
 'свадьба' 'получение дополнительного образования' 'покупка своего жилья'
 'операции с недвижимостью' 'получение высшего образования'
 'свой автомобиль' 'сделка с автомобилем' 'профильное образование'
 'высшее образование' 'покупка жилья для сдачи' 'на покупку автомобиля'
 'ремонт жилью' 'заняться высшим образованием']


In [41]:
lemmas = []
for p in unique_purposes: #для каждой цели в уникальных целях получения кредита
    lemmas += m.lemmatize(p) #добавляем лемматизированную строку в список

In [42]:
print(Counter(lemmas))

Counter({' ': 59, '\n': 38, 'покупка': 10, 'недвижимость': 10, 'автомобиль': 9, 'образование': 9, 'жилье': 7, 'с': 5, 'операция': 4, 'на': 4, 'свой': 4, 'свадьба': 3, 'строительство': 3, 'получение': 3, 'высокий': 3, 'дополнительный': 2, 'для': 2, 'коммерческий': 2, 'жилой': 2, 'заниматься': 2, 'сделка': 2, 'приобретение': 1, 'сыграть': 1, 'проведение': 1, 'семья': 1, 'собственный': 1, 'подержать': 1, 'со': 1, 'подержанный': 1, 'профильный': 1, 'сдача': 1, 'ремонт': 1})


Из лемм становятся ясны основные цели получения кредита.

Подробнее рассмотрим "недвижимость", "жилье", а также "ремонт" (кредит на ремонт чего? не пересекаются ли слова?).

In [43]:
lemmas_detailed = []
for string in unique_purposes:
    lemmatized = m.lemmatize(string)
    if (('ремонт' in lemmatized) or \
        ('недвижимость' in lemmatized) or \
        ('жилье' in lemmatized) or \
        ('операция' in lemmatized)) & (lemmatized not in lemmas_detailed):
        lemmas_detailed.append(lemmatized)

In [44]:
for lemma_purpose in lemmas_detailed:
    print(lemma_purpose)

['покупка', ' ', 'жилье', '\n']
['операция', ' ', 'с', ' ', 'жилье', '\n']
['покупка', ' ', 'жилье', ' ', 'для', ' ', 'семья', '\n']
['покупка', ' ', 'недвижимость', '\n']
['покупка', ' ', 'коммерческий', ' ', 'недвижимость', '\n']
['покупка', ' ', 'жилой', ' ', 'недвижимость', '\n']
['строительство', ' ', 'собственный', ' ', 'недвижимость', '\n']
['недвижимость', '\n']
['строительство', ' ', 'недвижимость', '\n']
['операция', ' ', 'с', ' ', 'коммерческий', ' ', 'недвижимость', '\n']
['строительство', ' ', 'жилой', ' ', 'недвижимость', '\n']
['жилье', '\n']
['операция', ' ', 'со', ' ', 'свой', ' ', 'недвижимость', '\n']
['покупка', ' ', 'свой', ' ', 'жилье', '\n']
['операция', ' ', 'с', ' ', 'недвижимость', '\n']
['покупка', ' ', 'жилье', ' ', 'для', ' ', 'сдача', '\n']
['ремонт', ' ', 'жилье', '\n']


Объединим причины с леммами "недвижимость" и "жилье": жилье является недвижимостью, кроме того, это понятие растяжимое.

Ремонт является отдельной причиной получения кредита и будет выделен в собственную категорию.

In [45]:
#создадим столбец с краткими причинами получения кредита
def purpose_compactize(purpose):
    if 'ремонт' in purpose:
        return 'жилищный ремонт' #ремонт будет первым условием, чтобы строки с ним не были отнесены к покупке недвижимости
    elif 'недвижимост' in purpose or 'жиль' in purpose:
        return 'приобретение недвижимости'
    elif 'автомобил' in purpose:
        return 'приобретение автомобиля'
    elif 'образовани' in purpose:
        return 'оплата образовательных услуг'
    elif 'свадьб' in purpose:
        return 'свадьба'

In [46]:
data['compact_purpose'] = data['purpose'].apply(purpose_compactize)

Проверим наличие пропусков в `compact_purpose`:

In [47]:
data['compact_purpose'].isna().sum()

0

Проверим, какие уникальные значения записаны в `compact_purpose`:

In [48]:
data['compact_purpose'].unique()

array(['приобретение недвижимости', 'приобретение автомобиля',
       'оплата образовательных услуг', 'свадьба', 'жилищный ремонт'],
      dtype=object)

Проверим соответствие `purpose` и `compact_purpose`:

In [49]:
data[['purpose', 'compact_purpose']].head(10) #проверяем соответствие

purpose               compact_purpose
0               покупка жилья     приобретение недвижимости
1     приобретение автомобиля       приобретение автомобиля
2               покупка жилья     приобретение недвижимости
3  дополнительное образование  оплата образовательных услуг
4             сыграть свадьбу                       свадьба
5               покупка жилья     приобретение недвижимости
6           операции с жильем     приобретение недвижимости
7                 образование  оплата образовательных услуг
8       на проведение свадьбы                       свадьба
9     покупка жилья для семьи     приобретение недвижимости

Проверим, соответствие значений для кредита на ремонт:

In [50]:
data.loc[data['compact_purpose'] == 'жилищный ремонт', ['purpose', 'compact_purpose']].head(5)

purpose  compact_purpose
139  ремонт жилью  жилищный ремонт
163  ремонт жилью  жилищный ремонт
228  ремонт жилью  жилищный ремонт
270  ремонт жилью  жилищный ремонт
276  ремонт жилью  жилищный ремонт

**Вывод**

Для систематизации целей получения кредита значения столбца purpose были лемматизированы с помощью средств библиотеки pymystem3. На основании часто встречающихся лемм, посчитанных в контейнере Counter, были выделены 5 кратких категорий целей получения кредита, а с помощью функции, примененной к purpose методом apply, для них создан отдельный столбец с соответствующими целями.

### Категоризация данных

Назначим для признаков, записанных в датафрейме, категории в числовом выражении, в словарях запишем соответствие числа и значения.

In [51]:
compact_purposes = data['compact_purpose'].unique()
purpose_dict = pd.DataFrame({'purpose_id': range(5), 'compact_purpose': compact_purposes})

In [52]:
purpose_dict

purpose_id               compact_purpose
0           0     приобретение недвижимости
1           1       приобретение автомобиля
2           2  оплата образовательных услуг
3           3                       свадьба
4           4               жилищный ремонт

Добавляем id цели получения кредита в датафрейм:

In [53]:
data = data.merge(purpose_dict, on='compact_purpose', how='left') 

In [54]:
data[['compact_purpose', 'purpose_id']].head(10) #проверяем соответствие

compact_purpose  purpose_id
0     приобретение недвижимости           0
1       приобретение автомобиля           1
2     приобретение недвижимости           0
3  оплата образовательных услуг           2
4                       свадьба           3
5     приобретение недвижимости           0
6     приобретение недвижимости           0
7  оплата образовательных услуг           2
8                       свадьба           3
9     приобретение недвижимости           0

In [55]:
data.loc[data['purpose_id'] == 4, 'compact_purpose'].head()

139    жилищный ремонт
163    жилищный ремонт
228    жилищный ремонт
270    жилищный ремонт
276    жилищный ремонт
Name: compact_purpose, dtype: object

Составим словари для каждого признака заемщика:

In [56]:
education_dict = data[['education_id', 'education']].drop_duplicates().reset_index(drop=True)
education_dict

education_id            education
0             0               высшее
1             1              среднее
2             2  неоконченное высшее
3             3            начальное
4             4       ученая степень

In [57]:
family_status_dict = data[['family_status_id', 'family_status']].drop_duplicates().reset_index(drop=True)
family_status_dict['family_status'] = family_status_dict['family_status'].str.lower() #нижний регистр
family_status_dict

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  не женат / не замужем

In [58]:
gender_dict = pd.DataFrame({'gender_id': [0, 1], 'gender': ['F', 'M']})
gender_dict

gender_id gender
0          0      F
1          1      M

In [59]:
data = data.merge(gender_dict, on='gender', how='left')

Проверим количество клиентов каждого типа занятости:

In [60]:
data['income_type'].value_counts()

сотрудник          11084
компаньон           5078
пенсионер           3829
госслужащий         1457
безработный            2
предприниматель        2
студент                1
в декрете              1
Name: income_type, dtype: int64

Необходимо распределить клиентов на более широкие категории типа занятости.

Сотрудник может быть госслужащим, это более широкая категория, сотрудница в декрете официально трудоустроена, но находится в отпуске, а предприниматель не может быть признан государством безработным.

"Напрашиваются" 3 категории: работающий, пенсионер, не трудоустроен:

In [61]:
employment_dict = pd.DataFrame({'employment_id': range(3), 'employment': ['работающий', 'пенсионер', 'не трудоустроен']})
employment_dict

employment_id       employment
0              0       работающий
1              1        пенсионер
2              2  не трудоустроен

In [62]:
def sort_employment(type_of_income): # функция, которая возвращает категорию занятости
    if type_of_income == 'компаньон' or type_of_income == 'безработный' or type_of_income == 'студент':
        return 'не трудоустроен'
    elif type_of_income == 'пенсионер':
        return type_of_income
    else:
        return 'работающий'

In [63]:
data['employment'] = data['income_type'].apply(sort_employment)

In [64]:
data = data.merge(employment_dict, on='employment', how='left')

Соответствие типа занятости новым столбцам:

In [65]:
data[['income_type', 'employment_id', 'employment']].head(10)

income_type  employment_id       employment
0   сотрудник              0       работающий
1   сотрудник              0       работающий
2   сотрудник              0       работающий
3   сотрудник              0       работающий
4   пенсионер              1        пенсионер
5   компаньон              2  не трудоустроен
6   компаньон              2  не трудоустроен
7   сотрудник              0       работающий
8   сотрудник              0       работающий
9   сотрудник              0       работающий

25-й, 50-й и 75-й процентили для дней трудового стажа:

In [66]:
middle_employ_median = data.loc[:, 'days_employed'].median()
low_employ_25 = data.loc[:, 'days_employed'].quantile(q=0.25)
high_employ_75 = data.loc[:, 'days_employed'].quantile(q=0.75)
print(low_employ_25, '|', middle_employ_median, '|', high_employ_75)

1023.0 | 2194.0 | 4797.0


In [67]:
#для удобства переведем дни стажа в года
print(low_employ_25 / 365)
print(middle_employ_median / 365)
print(high_employ_75 / 365)

2.802739726027397
6.010958904109589
13.142465753424657


В соответствии со значениями квартилей составим категоризацию заемщиков по годам трудового стажа:

In [68]:
work_experience_dict = pd.DataFrame({'work_experience_id': range(4), 
                                     'work_experience': ['менее 2.8 лет', '2.8-6 лет', '>6-13.2 лет', 'более 13.2 лет']})

In [69]:
work_experience_dict

work_experience_id work_experience
0                   0   менее 2.8 лет
1                   1       2.8-6 лет
2                   2     >6-13.2 лет
3                   3  более 13.2 лет

In [70]:
def sort_experience(days): # функция для присвоения категории трудового стажа
    try:
        if days < (2.8 * 365):
            return 0
        elif (2.8 * 365) <= days <= (6 * 365):
            return 1
        elif (6 * 365) < days <= (13.2 * 365):
            return 2
        else:
            return 3
    except:
        print('значения нет')
        
data['work_experience_id'] = data['days_employed'].apply(sort_experience)

In [71]:
data[['work_experience_id', 'days_employed']].head()

work_experience_id  days_employed
0                   3           8437
1                   2           4024
2                   3           5623
3                   2           4124
4                   3          14177

Составим категории для дохода. Вычислим квартили и на их основе создадим словарь категоризации дохода заемщиков.

In [72]:
middle_income_median = data['total_income'].median()
low_income_25 = data.loc[:, 'total_income'].quantile(q=0.25)
high_income_75 = data.loc[:, 'total_income'].quantile(q=0.75)
print('\tКвартили для дохода:\n', low_income_25, '|', middle_income_median, '|', high_income_75)

	Квартили для дохода:
 107623.0 | 142594.0 | 195820.25


In [73]:
income_dict = pd.DataFrame({'income_id': range(4), 
                            'income': ['lowest', 'lower middle', 'higher middle', 'highest']})

In [74]:
income_dict

income_id         income
0          0         lowest
1          1   lower middle
2          2  higher middle
3          3        highest

In [75]:
def sort_income(income): # функция для категоризации по уровню ежемесячного дохода
    try:
        if income < low_income_25:
            return 'lowest'
        elif low_income_25 <= income <= middle_income_median:
            return 'lower middle'
        elif middle_income_median < income < high_income_75:
            return 'higher middle'
        else:
            return 'highest'
    except:
        print('значения нет')
        
data['income'] = data['total_income'].apply(sort_income)
data = data.merge(income_dict, on='income', how='left')

In [76]:
data[['total_income', 'income_id', 'income']].head()

total_income  income_id         income
0        253875          3        highest
1        112080          1   lower middle
2        145885          2  higher middle
3        267628          3        highest
4        158616          2  higher middle

Для категоризации возраста используем классификацию возрастов ВОЗ

In [77]:
age_dict = pd.DataFrame({'age_id': range(4),
                         'age': ['юный', 'молодой', 'зрелый', 'пенсионный']})

In [78]:
age_dict

age_id         age
0       0        юный
1       1     молодой
2       2      зрелый
3       3  пенсионный

In [79]:
def sort_age(age): # функция для категоризации заемщиков по возрасту
    try:
        if age < 30:
            return 0
        elif age < 45:
            return 1
        elif age < 60:
            return 2
        else:
            return 3
    except:
        print('значения нет')
        
data['age_id'] = data['dob_years'].apply(sort_age)

In [80]:
data[['age_id', 'dob_years']].merge(age_dict, how='left').tail(10) #проверка соответствия

age_id  dob_years         age
21444       0         28        юный
21445       1         42     молодой
21446       1         42     молодой
21447       2         59      зрелый
21448       1         37     молодой
21449       1         43     молодой
21450       3         67  пенсионный
21451       1         38     молодой
21452       1         38     молодой
21453       1         40     молодой

Так выглядит датафрейм из категорий без строковых значений, расшифровка для всех значений записана в соответствующих словарях:

In [81]:
data[['children', 
      'family_status_id', 
      'employment_id', 
      'work_experience_id', 
      'income_id', 
      'age_id',
      'gender_id', 
      'education_id', 
      'purpose_id',
      'debt']].head(5) 

children  family_status_id  employment_id  work_experience_id  income_id  \
0         1                 0              0                   3          3   
1         1                 0              0                   2          1   
2         0                 0              0                   3          2   
3         3                 0              0                   2          3   
4         0                 1              1                   3          2   

   age_id  gender_id  education_id  purpose_id  debt  
0       1          0             0           0     0  
1       1          0             1           1     0  
2       1          1             1           0     0  
3       1          1             1           2     0  
4       2          0             1           3     0

### Вывод

Строковые значения усложняют работу с данными, в связи с чем для каждого столбца со строковыми значениями создан столбец с соответствующими id. Для каждой категории создан словарь.

Доходы и стаж категоризированы на основании разделения их на 4 равные части (стаж для удобства переведен в года), а возраст категоризирован на основании классификации возрастов ВОЗ.

Кроме того, были категоризированы типы занятости: распределены между 3 большими категориями (работающий, пенсионер и не трудоустроен).

## Связь присущих заемщикам признаков и возврата кредита в срок

### Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [82]:
children_debt = data.groupby('children')['debt'].agg({'sum', 'count'})
children_debt = children_debt.rename(columns={'count': 'total_borrowers', 'sum': 'non_payment'})
children_debt['debt_ratio'] = children_debt['non_payment'] / children_debt['total_borrowers']

In [83]:
children_debt.sort_values('debt_ratio', ascending=False).style.format({'debt_ratio': '{:.2%}'})

#### Вывод

Взаимосвязь между наличием детей и возвратом кредита в срок есть. 

Бездетные клиенты возвращают кредит в срок чаще на 1.5 процентных пункта, чем клиенты, у которых есть хотя бы один ребенок. Хотя клиенты с 5 детьми ни разу не просрочили возврат кредита, на основании этого делать выводы не стоит - таких клиентов всего 9.

### Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [84]:
family_status_debt = data.groupby('family_status_id')['debt'].agg({'sum', 'count'})
family_status_debt = family_status_debt.rename(columns={'count': 'total_borrowers', 'sum': 'non_payment'})
family_status_debt['debt_ratio'] = family_status_debt['non_payment'] / family_status_debt['total_borrowers']

In [85]:
family_status_debt.sort_values('debt_ratio', ascending=False).merge(family_status_dict, on='family_status_id')\
                  .style.format({'debt_ratio': '{:.2%}'})

#### Вывод

Самый низкий индекс невозврата кредита имеют клиенты категории "вдовец / вдова", самый высокий индекс у категории "гражданский брак" и "не женат / не замужем". Просрочка по кредиту у них бывает чаще чем у клиентов когда-либо бывших или состоящих в официальном браке почти на 2 процентных пункта.

Таким образом, взаимосвязь между семейным положением и возвратом кредита в срок есть: реже всего кредит возвращают в срок клиенты, которые никогда не состояли в официально зарегистрированном браке.

### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [86]:
income_debt = data.groupby('income_id')['debt'].agg({'sum', 'count'})
income_debt = income_debt.rename(columns={'count': 'total_borrowers', 'sum': 'non_payment'})
income_debt['debt_ratio'] = income_debt['non_payment'] / income_debt['total_borrowers']

In [87]:
income_debt.sort_values('debt_ratio', ascending=False).merge(income_dict, on='income_id')\
           .style.format({'debt_ratio': '{:.2%}'})

#### Вывод

Зависимость между уровнем дохода и возвратом кредита в срок есть: как ни странно, в срок возвращают кредит чаще всего и клиенты с самыми высокими доходами, и клиенты с самыми низкими доходами. Возможно, дело в том, что клиенты с самыми низкими доходами берут кредит на небольшие суммы, которые легче вернуть.

Клиенты из категорий среднего дохода возвращают кредит в срок реже на 0.7-1 процентных пункта.

Таким образом, зависимость между уровнем дохода и возвратом кредита в срок есть: клиенты со средним доходом возвращают кредит в срок реже.

### Как разные цели кредита влияют на его возврат в срок?

In [88]:
purpose_debt = data.groupby('compact_purpose')['debt'].agg({'sum', 'count'})
purpose_debt = purpose_debt.rename(columns={'count': 'total_borrowers', 'sum': 'non_payment'})
purpose_debt['debt_ratio'] = purpose_debt['non_payment'] / purpose_debt['total_borrowers']

In [89]:
purpose_debt.sort_values('debt_ratio', ascending=False).style.format({'debt_ratio': '{:.2%}'})

#### Вывод

Чаще всего возвращают кредит в срок клиенты, которые брали его на жилищный ремонт - менее 6 процентных пунктов невозврата в срок.

Реже всего в срок возвращают кредит клиенты, которые брали его на автомобиль и на оплату образовательных услуг (что неудивительно, клиенты, получающиеся образование, наверняка нечасто имеют возможность трудоустроиться на полную ставку), они возвращают кредит в срок реже, чем кредиты на ремонт, на 3.5 процентных пункта.

## Общий вывод

В ходе анализа были изучены, обработаны и категоризированы данные, что в итоге дало возможность найти взаимосвязи между возвратом кредита в срок и наличием детей, семейным положением, уровнем дохода, целями кредита:

- заемщики, у которых нет детей, возвращают кредит в срок чаще на 1.5 процентных пункта, чем те, у кого есть хотя бы один ребенок;

- реже всего кредит возвращают в срок клиенты, которые никогда не состояли в официально зарегистрированном браке, клиенты со средним доходом и те, кто брал кредит на автомобиль и на оплату образовательных услуг.

Кроме того, благодаря категоризации, возможен дальнейший поиск взаимосвязей (например, связь с образованием, типом занятости), которые можно использовать для построения моделей.